File: android-lib/src/androidTest/java/com/kinvey/androidTest/cache/QueryTest.java
package com.kinvey.androidTest.cache;


import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.test.RenamingDelegatingContext;
import android.test.suitebuilder.annotation.SmallTest;

import com.kinvey.android.cache.QueryHelper;
import com.kinvey.java.Query;
import com.kinvey.java.query.AbstractQuery;
import com.kinvey.java.query.MongoQueryFilter;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import io.realm.DynamicRealm;
import io.realm.DynamicRealmObject;
import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmQuery;
import io.realm.Sort;

import static junit.framework.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by Prots on 30/01/16.
 */
@RunWith(AndroidJUnit4.class)
@SmallTest
public class QueryTest {

    private RealmQuery<DynamicRealmObject> query;
    private static final String TEST_FIELD = "field";

    @Before
    public void setup(){
        Context mMockContext = new RenamingDelegatingContext(InstrumentationRegistry.getInstrumentation().getTargetContext(), "test_");
        Realm.init(mMockContext);
        DynamicRealm realm = DynamicRealm.getInstance(new RealmConfiguration.Builder().build());

        if (!realm.getSchema().contains("test")){
            realm.beginTransaction();
            realm.getSchema().create("test").addField("_id", Integer.class);
            realm.commitTransaction();
        }

        RealmQuery query = mock(RealmQuery.class);
        doReturn(query).when(query).beginGroup();
        doReturn(query).when(query).equalTo(anyString(), anyString());
        doReturn(query).when(query).endGroup();
        doReturn(query).when(query).or();
        doReturn(query).when(query).not();
        this.query = query;
    }

    @After
    public void tearDown() {
        Context mMockContext = new RenamingDelegatingContext(InstrumentationRegistry.getInstrumentation().getTargetContext(), "test_");
        Realm.init(mMockContext);
        DynamicRealm realm = DynamicRealm.getInstance(new RealmConfiguration.Builder().build());
        if (realm.getSchema().contains("test")){
            realm.beginTransaction();
            realm.getSchema().remove("test");
            realm.commitTransaction();
        }
    }

    @Test
    public void testInClause(){
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.in("_id", new String[]{"1", "2"});

        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());

        verify(query, times(1)).beginGroup();
        verify(query, times(2)).equalTo(eq("_id"), anyString());
        verify(query, times(1)).or();
        verify(query, times(1)).endGroup();
    }

    @Test
    public void testNotInClause(){
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notIn("_id", new String[]{"1", "2"});

        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());

        verify(query, times(2)).beginGroup();
        verify(query, times(2)).equalTo(eq("_id"), anyString());
        verify(query, times(1)).or();
        verify(query, times(2)).endGroup();
        verify(query, times(1)).not();
    }

    @Test
    public void testGreaterThanIntClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThan(TEST_FIELD, 1);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThan(TEST_FIELD, 1);
    }

    @Test
    public void testGreaterThanLongClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThan(TEST_FIELD, 1L);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThan(TEST_FIELD, 1L);
    }

    @Test
    public void testGreaterThanDoubleClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThan(TEST_FIELD, 1.0d);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThan(TEST_FIELD, 1.0d);
    }

    @Test
    public void testGreaterThanFloatClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThan(TEST_FIELD, 1.0f);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThan(TEST_FIELD, 1.0f);
    }

    @Test
    public void testGreaterThanOrEqualToIntClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThanEqualTo(TEST_FIELD, 1);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThanOrEqualTo(TEST_FIELD, 1);
    }

    @Test
    public void testGreaterThanOrEqualToLongClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThanEqualTo(TEST_FIELD, 1L);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThanOrEqualTo(TEST_FIELD, 1L);
    }

    @Test
    public void testGreaterThanOrEqualToDoubleClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThanEqualTo(TEST_FIELD, 1.0d);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThanOrEqualTo(TEST_FIELD, 1.0d);
    }

    @Test
    public void testGreaterThanOrEqualToFloatClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.greaterThanEqualTo(TEST_FIELD, 1.0f);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).greaterThanOrEqualTo(TEST_FIELD, 1.0f);
    }


    @Test
    public void testLessThanIntClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThan(TEST_FIELD, 1);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThan(TEST_FIELD, 1);
    }

    @Test
    public void testLessThanLongClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThan(TEST_FIELD, 1L);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThan(TEST_FIELD, 1L);
    }

    @Test
    public void testLessThanDoubleClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThan(TEST_FIELD, 1.0d);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThan(TEST_FIELD, 1.0d);
    }

    @Test
    public void testLessThanFloatClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThan(TEST_FIELD, 1.0f);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThan(TEST_FIELD, 1.0f);
    }

    @Test
    public void testLessThanOrEqualToIntClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThanEqualTo(TEST_FIELD, 1);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThanOrEqualTo(TEST_FIELD, 1);
    }

    @Test
    public void testLessThanOrEqualToLongClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThanEqualTo(TEST_FIELD, 1L);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThanOrEqualTo(TEST_FIELD, 1L);
    }

    @Test
    public void testLessThanOrEqualToDoubleClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThanEqualTo(TEST_FIELD, 1.0d);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThanOrEqualTo(TEST_FIELD, 1.0d);
    }

    @Test
    public void testLessThanOrEqualToFloatClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.lessThanEqualTo(TEST_FIELD, 1.0f);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).lessThanOrEqualTo(TEST_FIELD, 1.0f);
    }

    @Test
    public void testNotEqualBooleanClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, true);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, true);
    }

    @Test
    public void testNotEqualByteClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        Byte aByte = 1;
        q.notEqual(TEST_FIELD, aByte);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, aByte);
    }

    @Test
    public void testNotEqualByteArrayClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        byte[] bytes = {1};
        q.notEqual(TEST_FIELD, bytes);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, bytes);
    }

    @Test
    public void testNotEqualShortClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        Short aShort = 12345;
        q.notEqual(TEST_FIELD, aShort);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, aShort);
    }

    @Test
    public void testNotEqualIntClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, 1);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, 1);
    }

    @Test
    public void testNotEqualLongClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, 1L);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, 1L);
    }

    @Test
    public void testNotEqualDoubleClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, 1d);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, 1d);
    }

    @Test
    public void testNotEqualFloatClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, 1f);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, 1f);
    }

    @Test
    public void testNotEqualStringClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, TEST_FIELD);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, TEST_FIELD);
    }

    @Test
    public void tesEqualByteClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        Byte aByte = 1;
        q.equals(TEST_FIELD, aByte);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, aByte);
    }

    @Test
    public void testEqualByteArrayClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        byte[] bytes = {1};
        q.equals(TEST_FIELD, bytes);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, bytes);
    }

    @Test
    public void testEqualShortClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        Short aShort = 12345;
        q.equals(TEST_FIELD, aShort);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, aShort);
    }

    @Test
    public void testEqualIntClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.equals(TEST_FIELD, 1);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, 1);
    }

    @Test
    public void testEqualLongClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.equals(TEST_FIELD, 1L);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, 1L);
    }

    @Test
    public void testEqualDoubleClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.equals(TEST_FIELD, 1d);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, 1d);
    }

    @Test
    public void testEqualFloatClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.equals(TEST_FIELD, 1f);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, 1f);
    }

    @Test
    public void testEqualStringClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.equals(TEST_FIELD, TEST_FIELD);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).equalTo(TEST_FIELD, TEST_FIELD);
    }

    @Test
    public void testAddSortASC() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.addSort(TEST_FIELD, AbstractQuery.SortOrder.ASC);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());

        //check that addSort isn't used in prepareRealmQuery
        verify(query, times(0)).beginGroup();
        verify(query, times(0)).equalTo(eq(TEST_FIELD), anyString());
        verify(query, times(0)).or();
        verify(query, times(0)).not();
        verify(query, times(0)).endGroup();

        assertEquals("{\"field\" : 1}", q.getSortString());
    }

    @Test
    public void testAddSortDESC() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.addSort(TEST_FIELD, AbstractQuery.SortOrder.DESC);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());

        //check that addSort isn't used in prepareRealmQuery
        verify(query, times(0)).beginGroup();
        verify(query, times(0)).equalTo(eq(TEST_FIELD), anyString());
        verify(query, times(0)).or();
        verify(query, times(0)).not();
        verify(query, times(0)).endGroup();

        assertEquals("{\"field\" : -1}", q.getSortString());
    }

    @Test
    public void testOrClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        Query q2 = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.in(TEST_FIELD, new String[]{"1", "2"});
        q2.in(TEST_FIELD, new String[]{"1", "2"});
        q.or(q2);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(5)).beginGroup();
        verify(query, times(4)).equalTo(eq(TEST_FIELD), anyString());
        verify(query, times(3)).or();
        verify(query, times(5)).endGroup();
    }

    @Test
    public void testAndClause() {
        Query q = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        Query q2 = new Query(new MongoQueryFilter.MongoQueryFilterBuilder());
        q.notEqual(TEST_FIELD, 1);
        q2.notEqual(TEST_FIELD, 2);
        q.and(q2);
        QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        verify(query, times(1)).notEqualTo(TEST_FIELD, 1);
        verify(query, times(1)).notEqualTo(TEST_FIELD, 2);
        verify(query, times(3)).beginGroup();
        verify(query, times(3)).endGroup();
    }

    @Test
    public void testUnsupportedOperationException() {
        Query q = new Query();
        q.startsWith(TEST_FIELD, 1);
        try {
            QueryHelper.prepareRealmQuery(query, q.getQueryFilterMap());
        } catch (UnsupportedOperationException e) {
            assertEquals("this query is not supported by cache", e.getMessage());
        }
    }
    
}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/cache/QueryTest.kt
package com.kinvey.androidTest.store.data.cache

import android.content.Context
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import com.kinvey.android.cache.QueryHelper.prepareRealmQuery
import com.kinvey.java.Query
import com.kinvey.java.query.AbstractQuery.SortOrder
import com.kinvey.java.query.MongoQueryFilter.MongoQueryFilterBuilder
import io.realm.DynamicRealm
import io.realm.DynamicRealmObject
import io.realm.Realm
import io.realm.RealmConfiguration.Builder
import io.realm.RealmQuery
import junit.framework.Assert.assertEquals
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.*

/**
 * Created by Prots on 30/01/16.
 */
@RunWith(AndroidJUnit4::class)
@SmallTest
class QueryTest {
    private var query: RealmQuery<DynamicRealmObject>? = null
    @Before
    fun setup() {
        val mMockContext = InstrumentationRegistry.getInstrumentation().targetContext
        Realm.init(mMockContext)
        val realm: DynamicRealm = DynamicRealm.getInstance(Builder().build())
        if (!realm.schema.contains("test")) {
            realm.beginTransaction()
            realm.schema.create("test").addField("_id", Int::class.java)
            realm.commitTransaction()
        }
        val query = mock(RealmQuery::class.java)
        doReturn(query).`when`(query).beginGroup()
        doReturn(query).`when`(query).equalTo(ArgumentMatchers.anyString(), ArgumentMatchers.anyString())
        doReturn(query).`when`(query).endGroup()
        doReturn(query).`when`(query).or()
        doReturn(query).`when`(query).not()
        this.query = query as RealmQuery<DynamicRealmObject>?
    }

    @After
    fun tearDown() {
        val mMockContext: Context? = InstrumentationRegistry.getInstrumentation().targetContext
        Realm.init(mMockContext)
        val realm: DynamicRealm = DynamicRealm.getInstance(Builder().build())
        if (realm.schema.contains("test")) {
            realm.beginTransaction()
            realm.schema.remove("test")
            realm.commitTransaction()
        }
    }

    @Test
    fun testInClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.`in`("_id", arrayOf("1", "2"))
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.beginGroup()
        verify(query, times(2))?.equalTo(ArgumentMatchers.eq("_id"), ArgumentMatchers.anyString())
        verify(query, times(1))?.or()
        verify(query, times(1))?.endGroup()
    }

    @Test
    fun testNotInClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.notIn("_id", arrayOf<String?>("1", "2"))
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(2))?.beginGroup()
        verify(query, times(2))?.equalTo(ArgumentMatchers.eq("_id"), ArgumentMatchers.anyString())
        verify(query, times(1))?.or()
        verify(query, times(2))?.endGroup()
        verify(query, times(1))?.not()
    }

    @Test
    fun testGreaterThanIntClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThan(TEST_FIELD, 1)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThan(TEST_FIELD, 1)
    }

    @Test
    fun testGreaterThanLongClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThan(TEST_FIELD, 1L)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThan(TEST_FIELD, 1L)
    }

    @Test
    fun testGreaterThanDoubleClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThan(TEST_FIELD, 1.0)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThan(TEST_FIELD, 1.0)
    }

    @Test
    fun testGreaterThanFloatClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThan(TEST_FIELD, 1.0f)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThan(TEST_FIELD, 1.0f)
    }

    @Test
    fun testGreaterThanOrEqualToIntClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThanEqualTo(TEST_FIELD, 1)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThanOrEqualTo(TEST_FIELD, 1)
    }

    @Test
    fun testGreaterThanOrEqualToLongClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThanEqualTo(TEST_FIELD, 1L)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThanOrEqualTo(TEST_FIELD, 1L)
    }

    @Test
    fun testGreaterThanOrEqualToDoubleClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThanEqualTo(TEST_FIELD, 1.0)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThanOrEqualTo(TEST_FIELD, 1.0)
    }

    @Test
    fun testGreaterThanOrEqualToFloatClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.greaterThanEqualTo(TEST_FIELD, 1.0f)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.greaterThanOrEqualTo(TEST_FIELD, 1.0f)
    }

    @Test
    fun testLessThanIntClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThan(TEST_FIELD, 1)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThan(TEST_FIELD, 1)
    }

    @Test
    fun testLessThanLongClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThan(TEST_FIELD, 1L)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThan(TEST_FIELD, 1L)
    }

    @Test
    fun testLessThanDoubleClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThan(TEST_FIELD, 1.0)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThan(TEST_FIELD, 1.0)
    }

    @Test
    fun testLessThanFloatClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThan(TEST_FIELD, 1.0f)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThan(TEST_FIELD, 1.0f)
    }

    @Test
    fun testLessThanOrEqualToIntClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThanEqualTo(TEST_FIELD, 1)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThanOrEqualTo(TEST_FIELD, 1)
    }

    @Test
    fun testLessThanOrEqualToLongClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThanEqualTo(TEST_FIELD, 1L)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThanOrEqualTo(TEST_FIELD, 1L)
    }

    @Test
    fun testLessThanOrEqualToDoubleClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThanEqualTo(TEST_FIELD, 1.0)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThanOrEqualTo(TEST_FIELD, 1.0)
    }

    @Test
    fun testLessThanOrEqualToFloatClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.lessThanEqualTo(TEST_FIELD, 1.0f)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.lessThanOrEqualTo(TEST_FIELD, 1.0f)
    }

    @Test
    fun testNotEqualBooleanClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.notEqual(TEST_FIELD, true)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, true)
    }

    @Test
    fun testNotEqualByteClause() {
        val q = Query(MongoQueryFilterBuilder())
        val aByte: Byte = 1
        q.notEqual(TEST_FIELD, aByte)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, aByte)
    }

    @Test
    fun testNotEqualByteArrayClause() {
        val q = Query(MongoQueryFilterBuilder())
        val bytes = byteArrayOf(1)
        q.notEqual(TEST_FIELD, bytes)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, bytes)
    }

    @Test
    fun testNotEqualShortClause() {
        val q = Query(MongoQueryFilterBuilder())
        val aShort: Short = 12345
        q.notEqual(TEST_FIELD, aShort)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, aShort)
    }

    @Test
    fun testNotEqualIntClause() {
        val q = Query(MongoQueryFilterBuilder())
        val intVal: Int = 1
        q.notEqual(TEST_FIELD, intVal)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, intVal)
    }

    @Test
    fun testNotEqualLongClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.notEqual(TEST_FIELD, 1L)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, 1L)
    }

    @Test
    fun testNotEqualDoubleClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.notEqual(TEST_FIELD, 1.0)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, 1.0)
    }

    @Test
    fun testNotEqualFloatClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.notEqual(TEST_FIELD, 1f)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, 1f)
    }

    @Test
    fun testNotEqualStringClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.notEqual(TEST_FIELD, TEST_FIELD)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, TEST_FIELD)
    }

    @Test
    fun tesEqualByteClause() {
        val q = Query(MongoQueryFilterBuilder())
        val aByte: Byte = 1
        q.equals(TEST_FIELD, aByte)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, aByte)
    }

    @Test
    fun testEqualByteArrayClause() {
        val q = Query(MongoQueryFilterBuilder())
        val bytes = byteArrayOf(1)
        q.equals(TEST_FIELD, bytes)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, bytes)
    }

    @Test
    fun testEqualShortClause() {
        val q = Query(MongoQueryFilterBuilder())
        val aShort: Short = 12345
        q.equals(TEST_FIELD, aShort)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, aShort)
    }

    @Test
    fun testEqualIntClause() {
        val q = Query(MongoQueryFilterBuilder())
        val intVal: Int = 1
        q.equals(TEST_FIELD, intVal)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, intVal)
    }

    @Test
    fun testEqualLongClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.equals(TEST_FIELD, 1L)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, 1L)
    }

    @Test
    fun testEqualDoubleClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.equals(TEST_FIELD, 1.0)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, 1.0)
    }

    @Test
    fun testEqualFloatClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.equals(TEST_FIELD, 1f)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, 1f)
    }

    @Test
    fun testEqualStringClause() {
        val q = Query(MongoQueryFilterBuilder())
        q.equals(TEST_FIELD, TEST_FIELD)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.equalTo(TEST_FIELD, TEST_FIELD)
    }

    @Test
    fun testAddSortASC() {
        val q = Query(MongoQueryFilterBuilder())
        q.addSort(TEST_FIELD, SortOrder.ASC)
        prepareRealmQuery(query!!, q.queryFilterMap)

        //check that addSort isn't used in prepareRealmQuery
        verify(query, times(0))?.beginGroup()
        verify(query, times(0))?.equalTo(ArgumentMatchers.eq(TEST_FIELD), ArgumentMatchers.anyString())
        verify(query, times(0))?.or()
        verify(query, times(0))?.not()
        verify(query, times(0))?.endGroup()
        assertEquals("{\"field\" : 1}", q.sortString)
    }

    @Test
    fun testAddSortDESC() {
        val q = Query(MongoQueryFilterBuilder())
        q.addSort(TEST_FIELD, SortOrder.DESC)
        prepareRealmQuery(query!!, q.queryFilterMap)

        //check that addSort isn't used in prepareRealmQuery
        verify(query, times(0))?.beginGroup()
        verify(query, times(0))?.equalTo(ArgumentMatchers.eq(TEST_FIELD), ArgumentMatchers.anyString())
        verify(query, times(0))?.or()
        verify(query, times(0))?.not()
        verify(query, times(0))?.endGroup()
        assertEquals("{\"field\" : -1}", q.sortString)
    }

    @Test
    fun testOrClause() {
        val q = Query(MongoQueryFilterBuilder())
        val q2 = Query(MongoQueryFilterBuilder())
        q.`in`(TEST_FIELD, arrayOf("1", "2"))
        q2.`in`(TEST_FIELD, arrayOf("1", "2"))
        q.or(q2)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(5))?.beginGroup()
        verify(query, times(4))?.equalTo(ArgumentMatchers.eq(TEST_FIELD), ArgumentMatchers.anyString())
        verify(query, times(3))?.or()
        verify(query, times(5))?.endGroup()
    }

    @Test
    fun testAndClause() {
        val q = Query(MongoQueryFilterBuilder())
        val q2 = Query(MongoQueryFilterBuilder())
        val intVal1 = 1
        val intVal2 = 2
        q.notEqual(TEST_FIELD, intVal1)
        q2.notEqual(TEST_FIELD, intVal2)
        q.and(q2)
        prepareRealmQuery(query!!, q.queryFilterMap)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, intVal1)
        verify(query, times(1))?.notEqualTo(TEST_FIELD, intVal2)
        verify(query, times(3))?.beginGroup()
        verify(query, times(3))?.endGroup()
    }

    @Test
    fun testUnsupportedOperationException() {
        val q = Query()
        q.startsWith(TEST_FIELD, 1)
        try {
            prepareRealmQuery(query!!, q.queryFilterMap)
        } catch (e: UnsupportedOperationException) {
            assertEquals("this query is not supported by cache", e.message)
        }
    }

    companion object {
        private const val TEST_FIELD = "field"
    }
}